clang/lib/Driver/ToolChains/HLSL.cpp

```diff
@@ -20,6 +20,88 @@ using namespace llvm::opt;
 using namespace llvm;
 
 namespace {
+
+const unsigned OfflineLibMinor = 0xF;
+const unsigned MaxShaderModel6Minor = 7;
+bool isLegalShaderModel(Triple &T) {
+  if (T.getOS() != Triple::OSType::ShaderModel)
+    return false;
+
+  auto Kind = T.getEnvironment();
+
+  switch (Kind) {
+  default:
+    return false;
+  case Triple::EnvironmentType::Vertex:
+  case Triple::EnvironmentType::Hull:
+  case Triple::EnvironmentType::Domain:
+  case Triple::EnvironmentType::Geometry:
+  case Triple::EnvironmentType::Pixel:
+  case Triple::EnvironmentType::Compute:
+  case Triple::EnvironmentType::Library:
+  case Triple::EnvironmentType::Amplification:
+  case Triple::EnvironmentType::Mesh:
+    break;
+  }
+
+  auto Version = T.getOSVersion();
+  if (Version.getBuild())
+    return false;
+  if (Version.getSubminor())
+    return false;
+
+  auto OMinor = Version.getMinor();
+  if (!OMinor.hasValue())
+    return false;
+
+  unsigned Minor = OMinor.getValue();
+  unsigned Major = Version.getMajor();
+
+  switch (Major) {
```

Comment:
This code is really hard to follow. Switching on the shader stage and using version tuple compares makes more sense to me. Something like:

```
switch(Kind) {
  case Triple::EnvironmentType::Library:
  return Version >= VersionTuple(6,3);
  ...
  }
```